feature/project

Ahmed7osny1 opened the pull request.

Abdulrahman-Ragab-01 committed:
implement project usecases and its unit tests

Ahmed7osny1 committed:
update usecase and unit test

mosayed01 commented on src/main/kotlin/logic/usecase/project/CreateProjectUseCase.kt:
```diff
@@ -0,0 +1,32 @@
+import com.berlin.logic.generateIdHelper.DefaultIdGenerator
+import com.berlin.logic.repositories.ProjectRepository
+import com.berlin.model.Project
+import jdk.jfr.Description
```
Fot what usage this import?

Ahmed7osny1 replied:
this is mistake import

mosayed01 commented on src/test/kotlin/logic/usecase/project/CreateProjectUseCaseTest.kt:
```diff
@@ -0,0 +1,83 @@
+        )
+
+        // Then
+        result.onFailure { exception ->
```
Should the test fail when the result is success

Ahmed7osny1 replied:
Yes, this test should fail if the result is a success

Ahmed7osny1 replied:
the best use exceptionOrNull ?

mosayed01 replied:
U can check with the result class is equal to failure

Ahmed7osny1 replied:
return projectRepository.createProject(newProject)
                    .map { "Creation Successfully" }
                    .recover { "Creation Failed" }
                    
                    recover is applied only when the operation fails (results in Result.failure), that's right?

mosayed01 reviewed (dismissed):
Resolve comments, then it'll be fine to me

Ahmed7osny1 committed:
change naming of id generator class, and these parameters
solve a problem in creation state (change depend class on implementation to abstraction)

AhmadKharfan commented on src/main/kotlin/logic/usecase/project/UpdateProjectUseCase.kt:
```diff
@@ -0,0 +1,21 @@
+class UpdateProjectUseCase (
+    private val projectRepository: ProjectRepository
+) {
+    fun updateProject(project: Project): Result<String> {
```
i think you shouldn't pass the all project for update.
you must pass some property and make it nullable because the user may be update in one field and some property like id you shouldn't update it.

Ahmed7osny1 replied:
sure, i make some property nullable

Ahmed7osny1 committed:
Merge branch 'development' into feature/project

Ahmed7osny1 committed:
fix id generator error
